[PyTorch Mobile] Support torchbind custom classes in lite interpreter

iseeyuan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#51432 [PyTorch Mobile] Support torchbind custom classes in lite interpreter**
ghstack-source-id: 120761262

torchbind is a convenient way to include custom class to both python and torchscript. CREATE_OBJECT is used to create an object of custom class. 

CREATE_OBJECT was not supported by lite interpreter. The major reason was that for custom class directly defined in Python, there's no language parser in lite interpreter. It's still the case. However, for torchbind classes that are defined in C++, a python/torchscript parser is not needed.

This diff is to support the case of torchbind custom classes. 
1. The class type can be resolved at import level. 
2. If the class is not the supported torchbind class, an error message is provided at export stage. Workaround is also suggested. 
3. Unit tests. C++: ```LiteInterpreterTest::BuiltinClass``` is added as an end-to-end test on supported class. Python: ```test_unsupported_createobject``` is changed to ```test_unsupported_classtype``` to test unsupported classes.

Differential Revision: [D26168913](https://our.internmc.facebook.com/intern/diff/D26168913/)

iseeyuan committed:
[PyTorch Mobile] Support torchbind custom classes in lite interpreter



Differential Revision: [D26168913](https://our.internmc.facebook.com/intern/diff/D26168913/)

[ghstack-poisoned]

iseeyuan committed:
Update on "[PyTorch Mobile] Support torchbind custom classes in lite interpreter"


Differential Revision: [D26168913](https://our.internmc.facebook.com/intern/diff/D26168913/)

[ghstack-poisoned]

iseeyuan committed:
Update on "[PyTorch Mobile] Support torchbind custom classes in lite interpreter"


Differential Revision: [D26168913](https://our.internmc.facebook.com/intern/diff/D26168913/)

[ghstack-poisoned]

raziel commented on torch/csrc/jit/mobile/import.cpp:
```diff
@@ -216,8 +216,14 @@ void parseMethods(
+      }
     }
 
     function->set_register_size(register_size);
```
I guess, in the end, the check for "torch" in the type_resolver did not need to be added?

iseeyuan replied:
That would be the target. It's split to the second PR (will upload it soon). I'm looking for a way to not adding the "torch" type check.

iseeyuan commented on torch/csrc/jit/mobile/function.cpp:
```diff
@@ -58,7 +53,7 @@ bool Function::append_operator(
   }
 
-  if (model_version == 0x3L &&
+  if (model_version == (int64_t)0x3L &&
```
Good point. I searched a lot of static_cast in the code base. Will changed to static_cast.

iseeyuan commented on test/cpp/jit/test_lite_interpreter.cpp:
```diff
@@ -337,6 +338,87 @@ class TorchBindLiteInterpreterTestStruct
+
+    def forward(self, x) -> str:
+      return self.my_obj.get(x)
+  )",
```
It turns out to be written by arc lint. A little strange but I'll leave it to avoid the warnings.

iseeyuan committed:
Update on "[PyTorch Mobile] Support torchbind custom classes in lite interpreter"

ghstack-source-id: 120761262

torchbind is a convenient way to include custom class to both python and torchscript. CREATE_OBJECT is used to create an object of custom class. 

CREATE_OBJECT was not supported by lite interpreter. The major reason was that for custom class directly defined in Python, there's no language parser in lite interpreter. It's still the case. However, for torchbind classes that are defined in C++, a python/torchscript parser is not needed.

This diff is to support the case of torchbind custom classes. 
1. The class type can be resolved at import level. 
2. If the class is not the supported torchbind class, an error message is provided at export stage. Workaround is also suggested. 
3. Unit tests. C++: ```LiteInterpreterTest::BuiltinClass``` is added as an end-to-end test on supported class. Python: ```test_unsupported_createobject``` is changed to ```test_unsupported_classtype``` to test unsupported classes.

Differential Revision: [D26168913](https://our.internmc.facebook.com/intern/diff/D26168913/)

[ghstack-poisoned]

iseeyuan committed:
Update on "[PyTorch Mobile] Support torchbind custom classes in lite interpreter"

ghstack-source-id: 120761262

torchbind is a convenient way to include custom class to both python and torchscript. CREATE_OBJECT is used to create an object of custom class. 

CREATE_OBJECT was not supported by lite interpreter. The major reason was that for custom class directly defined in Python, there's no language parser in lite interpreter. It's still the case. However, for torchbind classes that are defined in C++, a python/torchscript parser is not needed.

This diff is to support the case of torchbind custom classes. 
1. The class type can be resolved at import level. 
2. If the class is not the supported torchbind class, an error message is provided at export stage. Workaround is also suggested. 
3. Unit tests. C++: ```LiteInterpreterTest::BuiltinClass``` is added as an end-to-end test on supported class. Python: ```test_unsupported_createobject``` is changed to ```test_unsupported_classtype``` to test unsupported classes.

Differential Revision: [D26168913](https://our.internmc.facebook.com/intern/diff/D26168913/)

[ghstack-poisoned]
